instrumentation/rails/lib/opentelemetry/instrumentation/rails/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+# Copyright 2020 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module Rails
+      module Middlewares
+        # TracerMiddleware propagates context and instruments Rails requests
+        # by way of its middleware system
+        class TracerMiddleware
+          def initialize(app)
+            @app = app
+          end
+
+          def call(env)
+            extracted_context = OpenTelemetry.propagation.http.extract(env)
+            OpenTelemetry::Context.with_current(extracted_context) do
+              tracer.in_span(env['PATH_INFO'], attributes: request_span_attributes(env: env), kind: :server) do |request_span|
+                @app.call(env).tap do |status, headers, response|
+                  set_attributes_after_request(
+                    request_span,
+                    env: env,
+                    status: status,
+                    headers: headers,
+                    response: response
+                  )
+                end
+              end
+            end
+          ensure
+            OpenTelemetry::Trace.current_span(extracted_context).finish if extracted_context
+          end
+
+          private
+
+          def request_destination_info(env)
+            ::Rails.application.routes.router.recognize(ActionDispatch::Request.new(env)) { |_route, params| return params }
+          end
+
+          def set_attributes_after_request(span, env:, status:, headers:, response:)
+            span.status = OpenTelemetry::Trace::Status.http_to_status(status)
+            span.set_attribute('http.status_code', status)
+            span.set_attribute('http.status_text', ::Rack::Utils::HTTP_STATUS_CODES[status])
+
+            controller = env['action_controller.instance']
+            return unless controller
+
+            span.name = "#{controller.class.name}.#{controller.action_name}"
+            span.set_attribute('rails.controller', controller.class.name)
+            span.set_attribute('rails.action', controller.action_name)
+          end
+
+          def request_span_attributes(env:)
+            {
+              'http.method' => env['REQUEST_METHOD'],
+              'http.host' => env['HTTP_HOST'] || 'unknown',
```

Comment:
If the host header is empty or absent, we should instead set either:
1. `http.server_name` and `net.host.port` OR
2. `net.host.name` and `net.host.port`

Note that `http.server_name` is supposed to be determined from configuration rather than headers - in our (Shopify's) case, I think we'd use the Shop's primary domain name.

I would probably try to use the second option here, since the first one is app-dependent.